symengine/sets.h

```diff
@@ -199,5 +225,42 @@ inline RCP<const Set> interval(const RCP<const Number> &start,
         return finiteset({start});
     return emptyset();
 }
+
+// ! \return RCP<const Set>
+inline RCP<const Set> sym_union(const set_basic &in, bool solve = true)
+{
+    set_basic input;
+    if (solve == false && in.size() > 1)
+        return make_rcp<const Union>(in);
+    set_basic combined_FiniteSet;
+    for (auto it = in.begin(); it != in.end(); ++it) {
+        if (is_a<FiniteSet>(**it)) {
+            set_basic container;
+            const FiniteSet &other = static_cast<const FiniteSet &>(**it);
+            std::set_union(combined_FiniteSet.begin(), combined_FiniteSet.end(),
+                           other.container_.begin(), other.container_.end(),
+                           std::inserter(container, container.begin()),
+                           RCPBasicKeyLess{});
+            combined_FiniteSet = container;
+        } else if (is_a<UniversalSet>(**it)) {
+            return universalset();
+        } else if (not is_a<EmptySet>(**it)) {
+            input.insert(*it);
+        }
+    }
+    if (input.empty()) {
+        return finiteset(combined_FiniteSet);
+    } else if (input.size() == 1 && combined_FiniteSet.empty()) {
+        return rcp_dynamic_cast<const Set>(*input.begin());
+    }
+    // Now we rely on respective containers' own rules
+    // TODO: Improve it to O(log n)
+    RCP<const Set> combined_Rest = finiteset(combined_FiniteSet);
+    for (auto it = input.begin(); it != input.end(); ++it) {
+        RCP<const Set> other = rcp_dynamic_cast<const Set>(*it);
+        combined_Rest = combined_Rest->set_union(other);
+    }
```

Comment:
Move the above for loop to a new method `interval_union`. And at the end, interval union and `combined_Finiteset` can be merged together.